### PR TITLE
Add Exponential Backoff Retry mechanism for any execution to Mongo

### PR DESCRIPTION
**Pull request by ReluXingZeng**
This change includes:

1. Exponential Backoff at execute function in Mongo.java
2. Update the checkstyle based on required output
3. Update some of our previous changes for checkstyle
4. Update README.md

### CODE REVIEW

**Review by daracurran — APPROVED** [review]
LGTM

File: driver/src/main/com/mongodb/Mongo.java [hunk 1]
```diff
@@ -767,23 +767,67 @@ OperationExecutor createOperationExecutor() {
...
+                }
+            });
         } finally {
             binding.release();
         }
     }
 
+    <T> T doRetryWithExponentialBackoff(final BasicCallback<T> mainAttempt) {
+        int failure = 0;
+        int maxFailure = 6;
```
Comment (melody4ever) [inline driver/src/main/com/mongodb/Mongo.java:800]: At least move this to a static final in the class.

**Comment by ReluXingZeng** [comment]
Don't merge it - find a potential issue related to it, at certain step we actually allow an Exception (e.g. MongoCommandException), they should still be thrown.

File: driver/src/main/com/mongodb/Mongo.java (hunk 1 above)
Comment (ReluXingZeng) [reply driver/src/main/com/mongodb/Mongo.java:800]: I actually prefer putting our code at the same spot :(, I know generically they should be associated with class instead :)

**Review by melody4ever — APPROVED** [review]

File: driver/src/main/com/mongodb/Mongo.java [hunk 2]
```diff
@@ -801,6 +801,11 @@ public T execute() {
             try {
                 return mainAttempt.execute();
             } catch (RuntimeException mainException) {
+                // Certain Exception are expected and are handled at our side. Throw.
+                if (mainException instanceof MongoCommandException) {
```
Comment (ReluXingZeng) [inline driver/src/main/com/mongodb/Mongo.java]: Also need to deal with `NoSuchElementException`, will add it.

**Review by romanvanderkrogt — APPROVED** [review]
LGTM.

**Pull request merged**